aptos-move/aptos-vm/src/sharded_block_executor/sharded_aggregator_service.rs

```diff
@@ -0,0 +1,155 @@
+    value.transpose().map_err(anyhow::Error::msg).unwrap()
+}
+
+/// This class ensures that deltas can use all 128 bits without having to let go of the sign bit for
+/// cases where the delta is negative. That is, we don't have to use conversions to i128.
+/// However, it does not handle overflow and underflow. That is, it will indicate to the caller of
+/// the faulty logic with their usage of deltas.
+#[derive(Clone, Copy)]
+struct DeltaU128 {
+    delta: u128,
+    is_positive: bool,
+}
+
+impl DeltaU128 {
+    pub fn get_delta(minuend: u128, subtrahend: u128) -> Self {
+        if minuend >= subtrahend {
+            Self {
+                delta: minuend - subtrahend,
+                is_positive: true,
+            }
+        } else {
+            Self {
+                delta: subtrahend - minuend,
+                is_positive: false,
+            }
+        }
+    }
+
+    fn add_delta(self, other: u128) -> u128 {
+        if self.is_positive {
+            self.delta + other
+        } else {
+            other - self.delta
+        }
+    }
+}
+
+impl Default for DeltaU128 {
+    fn default() -> Self {
+        Self {
+            delta: 0,
+            is_positive: true,
+        }
+    }
+}
+
+impl ops::Add for DeltaU128 {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        if self.is_positive == rhs.is_positive {
+            Self {
+                delta: self.delta + rhs.delta,
+                is_positive: self.is_positive,
+            }
+        } else {
+            let (pos, neg) = if self.is_positive {
+                (self.delta, rhs.delta)
+            } else {
+                (rhs.delta, self.delta)
+            };
+            if pos >= neg {
+                Self {
+                    delta: pos - neg,
+                    is_positive: true,
+                }
+            } else {
+                Self {
+                    delta: neg - pos,
+                    is_positive: false,
+                }
+            }
+        }
+    }
+}
+
+pub fn aggregate_and_update_total_supply<S: StateView>(
+    sharded_output: &mut Vec<Vec<Vec<TransactionOutput>>>,
+    global_output: &mut [TransactionOutput],
+    state_view: &S,
+) {
+    let num_shards = sharded_output.len();
+    let num_rounds = sharded_output[0].len();
+    let total_supply_base_val: u128 = get_state_value(&TOTAL_SUPPLY_STATE_KEY, state_view).unwrap();
+
+    let mut aggr_total_supply_delta = vec![DeltaU128::default(); num_shards * num_rounds + 1];
+
+    // No need to parallelize this as the runtime is O(num_shards)
+    // TODO: Get this from the individual shards while getting 'sharded_output'
+    sharded_output
+        .iter()
+        .enumerate()
+        .for_each(|(shard_id, shard_output)| {
+            for (round, txn_outputs) in shard_output.iter().enumerate() {
+                for last_txn in txn_outputs.iter().rev() {
+                    if let Some(last_txn_total_supply) =
```

Comment:
This is potentially dangerous and unnecessary.. all txns should have total supply except last txn. If so, can you handle the last txn as a special case but in all other case `unwrap` for the total supply value instead of using the iterator?

Reply:
done